pkg/storegateway/series_refs.go

```diff
@@ -798,11 +878,135 @@ func (s *loadingSeriesChunkRefsSetIterator) Next() bool {
 
 	s.currentSet = nextSet
 	if s.skipChunks && cachedSeriesID.isSet() {
-		storeCachedSeriesForPostings(s.ctx, s.indexCache, s.tenantID, s.blockID, s.shard, cachedSeriesID, nextSet, s.logger)
+		storeCachedSeriesForPostings(ctx, s.indexCache, s.tenantID, s.blockID, s.shard, cachedSeriesID, nextSet, s.logger)
 	}
 	return true
 }
 
+// clampLastChunkLength checks the length of the last chunk in the last range of the last series.
+// If the length of that chunk is larger than the difference with the first chunk ref in metas
+// then the length is clamped at that difference.
+// clampLastChunkLength assumes that the chunks are sorted by their refs
+// (currently this is equivalent to also being sorted by their minTime)
+// clampLastChunkLength is a noop if metas or series is empty.
+func clampLastChunkLength(series []seriesChunkRefs, metas []chunks.Meta) {
+	if len(series) == 0 || len(metas) == 0 {
+		return
+	}
+	var (
+		lastSeriesRanges = series[len(series)-1].chunksRanges
+		lastRange        = lastSeriesRanges[len(lastSeriesRanges)-1]
+		lastSeriesChunk  = lastRange.refs[len(lastRange.refs)-1]
+		firstRef         = metas[0].Ref
+	)
+
+	if lastRange.segmentFile != uint32(chunkSegmentFile(firstRef)) {
+		return
+	}
+	diffWithNextChunk := chunkOffset(firstRef) - lastSeriesChunk.segFileOffset
+	if lastSeriesChunk.length > diffWithNextChunk {
+		lastRange.refs[len(lastRange.refs)-1].length = diffWithNextChunk
+	}
+}
+
+// partitionChunks creates a slice of seriesChunkRefsRange for each range of chunks within the same segment file.
+// It may also create more ranges if there are more chunks.
+// The partitioning here should be fairly static and not depend on the actual Series() request because
+// the resulting ranges may be used for caching, and we want our cache entries to be reusable between requests.
+func partitionChunks(chks []chunks.Meta) [][]chunks.Meta {
+	const (
+		chunksRangesPerSeries = 2
+		minChunksPerRange     = 2
```

Comment:
Why do we need this min? Why 2?

Reply:
this is for series with very few chunks. If a series exists only for a few hours and has 3-4 chunks, it's less overhead in chunks caching if we group all of those chunks together.

I was actually thinking of making this a higher number. Maybe 10? With a scrape interval of 15s you'd get 48 chunks per 24h block. Currently we split this into two ranges of 24 chunks each. Maybe having minimum of 24 chunks per range makes sense? I honestly don't know, but definitely meant to increase this from 2 (the reason it's 2 is because some of the tests broke with higher numbers and I didn't go to fix them at the time 😬 )

Reply:
Have you considered reversing the logic into "maxChunksPerRange"?

Reply:
Discussed on Slack. Let's keep it as is. Thanks for the detailed information shared!